test/fixtures/es-module-loaders/assertionless-json-import.mjs

```diff
@@ -1,17 +1,22 @@
 const DATA_URL_PATTERN = /^data:application\/json(?:[^,]*?)(;base64)?,([\s\S]*)$/;
 const JSON_URL_PATTERN = /\.json(\?[^#]*)?(#.*)?$/;
 
-export function resolve(url, context, next) {
-  // Mutation from resolve hook should be discarded.
-  context.importAssertions.type = 'whatever';
-  return next(url);
-}
+export async function resolve(url, context, next) {
+  const { importAssertions } = context;
 
-export function load(url, context, next) {
-  if (context.importAssertions.type == null &&
+  if (importAssertions.type == null &&
       (DATA_URL_PATTERN.test(url) || JSON_URL_PATTERN.test(url))) {
-    const { importAssertions } = context;
-    importAssertions.type = 'json';
+
+    // Mutation from resolve hook should be discarded.
+    context.importAssertions.type = 'whatever';
```

Comment:
If I do, the test fails; because the context isn’t cloned before it’s passed into `defaultLoad`, which is where the assertion is validated. Previously we had this loader define its own `load` hook to cause the validation to get skipped; would we prefer that?

Reply:
I don't think that's correct, for me the following makes the test pass:

```js
const DATA_URL_PATTERN = /^data:application\/json(?:[^,]*?)(;base64)?,([\s\S]*)$/;
const JSON_URL_PATTERN = /\.json(\?[^#]*)?(#.*)?$/;

export async function resolve(url, context, next) {
  const { importAssertions: {type} } = context;
  
    // Mutation from resolve hook should be discarded.
    context.importAssertions.type = 'whatever';

  if (type == null &&
      (DATA_URL_PATTERN.test(url) || JSON_URL_PATTERN.test(url))) {

    // Clean new import assertions object to ensure that this test isn't passing due to mutation
    const newImportAssertions = { type: 'json' };

    const result = await next(url, context);
    result.importAssertions = newImportAssertions;
    return result;
  }

  return next(url, context);
}
```

Reply:
Actually, thinking more about this, `resolve` does not receive a URL, it receives a specifier, so as it is written, it would have all sort of false positives and false negatives. A more correct version would look like this:

```js
const DATA_URL_PATTERN = /^data:application\/json(?:[^,]*?)(;base64)?,([\s\S]*)$/;
const JSON_URL_PATTERN = /^[^?]+\.json(\?[^#]*)?(#.*)?$/;

export async function resolve(specifier, context, next) {
  const { importAssertions: { type } } = context;
  
  // Mutation from resolve hook should be discarded.
  context.importAssertions.type = 'whatever';

  const result = await next(specifier, context);

  if (type == null &&
      (DATA_URL_PATTERN.test(result.url) || JSON_URL_PATTERN.test(result.url))) {
    result.importAssertions ??= {}
    result.importAssertions.type = 'json';
    return result;
  }

  return next(url, context);
}
```

Reply:
I’m trying to understand why that wasn’t working for me. I’ve been making the changes one-by-one to compare the differences from my version to yours, and the one that makes the difference is changing my `const result = await next(url, { ...context, importAssertions: newImportAssertions });` to `const result = await next(url, context);`. Why would this change make a difference?

Reply:
Could it be that you were modifying `importAssertions.type` before checking if it was nullish?

Reply:
I feel like we shouldn’t be defining `importAssertions.type` with a bad value before passing it into the next hook:

```js
  // Mutation from resolve hook should be discarded.
  context.importAssertions.type = 'whatever';

  const result = await next(specifier, context);
```

This happens to work when the next hook is `defaultResolve`, because that ignores import assertions, but if another user loader happened to be chained after this one and it did something with the import assertions it would choke.

Reply:
IMO this test should be validating that defaultResolve does not care about the import assertions, and that's what this mutation is doing.

Reply:
> IMO this test should be validating that defaultResolve does not care about the import assertions

I was validating that `getModuleJob` didn’t care about the mutation. I thought that’s what the goal was here. We could probably do both, I guess? Though I’m starting to wonder if this object mutation stuff should get its own test so that it’s explicit what we’re trying to test for.

Reply:
Before this PR, it was modified before calling `next`, validating that defaultResolve did not care about it and the mutation had no effect on the rest of the process. If you want to remove that test, you should either replace it with something similar, or give a reason why we don't care about it.

>  It never even references context.importAssertions in its function source

Tests should not be written with knowledge of the source code, this is not a valid argument IMO.

Reply:
I think what was making it not work for me was that moving it up top would cause the `importAssertions.type == null` part to fail, because `type` was `whatever` by then. I revised to a version that’s more explicit about when and what we’re comparing.